File: 백준/Silver/4963. 섬의 개수/섬의 개수.java
import javax.swing.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static boolean[][] visited;
    public static int[] dx = {0,0,1,-1,1,1,-1,-1};
    public static int[] dy = {1,-1,0,0,1,-1,-1,1};
    public static void bfs(int[][] map , int[] p){
        ArrayDeque<int[]> q = new ArrayDeque<>();
        q.add(p);
        visited[p[0]][p[1]] = true;
        while(!q.isEmpty()){
            int[]cur = q.poll();
            int x = cur[1];
            int y = cur[0];
            for(int i = 0 ; i < 8 ; i++){
                int curX = x+dx[i];
                int curY = y+dy[i];
                if(curX < 0 || curY < 0 || curX >= map[0].length || curY >= map.length || visited[curY][curX])
                    continue;
                if(map[curY][curX] == 1 ) {
                    q.add(new int[]{curY, curX});
                    visited[curY][curX] = true;
                }
            }

        }

    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        String[] input;
        String[] line;
        while(true) {
            input = br.readLine().split(" ");
            int w = Integer.parseInt(input[0]);
            int h = Integer.parseInt(input[1]);
            if(w==0 && h ==0)
                break;

            int[][]map = new int[h][w];
            visited = new boolean[h][w];
            for(int i = 0 ; i < h ; i++){
                line = br.readLine().split(" ");
                for(int j = 0 ; j < w; j++){
                    map[i][j] = Integer.parseInt(line[j]);
                }
            }

            int cnt = 0;
            for(int i = 0 ; i < map.length ; i++){
                for(int j = 0 ; j < map[i].length ; j++){
                    if(map[i][j] == 1 && !visited[i][j]) {
                        bfs(map,new int[]{i,j});
                        cnt++;
                    }
                }
            }
            sb.append(cnt).append("\n");
        }
        System.out.println(sb);
    }
}

File: 백준/Silver/4963. 섬의 개수/섬의 개수.kt
package com.example.myapplication

import java.util.Stack

var visited:Array<BooleanArray>? = null
val dx = intArrayOf(0, 0, 1, -1, 1, 1, -1, -1)
val dy = intArrayOf(1, -1, 0, 0, 1, -1, -1, 1)
fun main(){
    val sb = StringBuilder()
    while(true){
        val (w,h) = readln().split(" ").map { it.toInt() }

        if(w==0 && h==0)
            break;

        val map = Array<IntArray>(h){IntArray(w)};
        visited = Array(h){BooleanArray(w)}

        for(i in 0 until h){
            val row = readln().split(" ").map{it.toInt()}
            for(j in 0 until w){
                map[i][j] = row[j]
            }
        }
        var cnt = 0
        for(i in 0 until h){
            for(j in 0 until w){
                if(map[i][j] == 1 && !visited!![i][j]){ // visited!! : visted를 non-nullable로 타입 바꿈
                    dfs(map,i,j);
                    cnt++
                }
            }
        }
        sb.append(cnt).append("\n")
    }
    print(sb)
}

fun dfs(map: Array<IntArray>, startY:Int, startX:Int){
    val s = Stack<IntArray>()
    s.push(intArrayOf(startY,startX))
    visited!![startY][startX] = true

    while(!s.isEmpty()){
        var cur = s.pop()
        var x = cur[1]
        var y = cur[0]

        for(i in 0 until 8){
            val curX = x+dx[i]
            val curY = y+dy[i]

            if (curX < 0 || curY < 0 || curX >= map[0].size || curY >= map.size || visited!![curY][curX]) continue
            if (map[curY][curX] == 1) {
                s.push(intArrayOf(curY, curX))
                visited!![curY][curX] = true
            }
        }
    }

}